server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyRequestBody.scala

```diff
@@ -36,7 +39,44 @@ class NettyRequestBody[F[_]](createFile: ServerRequest => F[TapirFile])(implicit
             Files.write(file.toPath, requestContentAsByteArray)
             RawValue(FileRange(file), Seq(FileRange(file)))
           })
-      case _: RawBodyType.MultipartBody => ???
+      case m: RawBodyType.MultipartBody =>
+        new SequenceSupport[F]()
+          .sequence[Part[Any]](getParts(serverRequest, m))
+          .map(RawValue.fromParts(_))
+    }
+  }
+
+  private def getParts(serverRequest: ServerRequest, m: RawBodyType.MultipartBody): List[F[Part[Any]]] = {
+    new HttpPostMultipartRequestDecoder(nettyRequest(serverRequest)).getBodyHttpDatas.asScala
+      .flatMap(httpData =>
+        httpData.getHttpDataType match {
+          case HttpDataType.Attribute =>
+            val part: F[Part[Any]] = monadError.unit(Part(name = httpData.getName, body = httpData.asInstanceOf[Attribute].getValue))
+            List(part)
+          case HttpDataType.FileUpload =>
+            m.partType(httpData.getName).map(c => handleNettyFileUpload(serverRequest, c, httpData.asInstanceOf[FileUpload])).toList
+          case HttpDataType.InternalAttribute => throw new UnsupportedOperationException("DataType not supported")
+        }
+      )
+      .toList
+  }
+
+  private def handleNettyFileUpload(serverRequest: ServerRequest, m: RawBodyType[_], upload: FileUpload): F[Part[Any]] = {
+    m match {
+      case RawBodyType.ByteArrayBody =>
+        monadError.unit(Part(name = upload.getName, body = upload.get(), contentType = MediaType.parse(upload.getContentType).toOption))
+      case RawBodyType.FileBody =>
+        createFile(serverRequest)
+          .map(file => {
+            Files.write(file.toPath, ByteBufUtil.getBytes(upload.content()))
```

Comment:
doesn't this read the whole file content into memory?

Reply:
I think yes

Reply:
well that's not good then. We need to transfer these bytes without reading the content into memory; and if there's any blocking code, surround this with `monadError.blocking`

Reply:
As far as I understand we could extract the package-private method `readBytes` from netty (https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/ByteBufUtil.java#L1878) and use it here so, that bytes from ByteBuffer could be written to `FileOutputStream`, depending on the size it is done directly or chunk by chunk.